politeiad/backendv2/tstorebe/plugins/pi/hooks.go

```diff
@@ -390,11 +392,256 @@ func (p *piPlugin) voteSummary(token []byte) (*ticketvote.SummaryReply, error) {
+		return true
+	}
+	// Convert comments slice to a map
+	commentsMap := make(map[uint32]comments.Comment, len(cs))
+	for _, c := range cs {
+		commentsMap[c.CommentID] = c
+	}
+
+	// Travel the comment tree searching for rootID.
+	// If we reach the tree haed without visiting the provided
+	// rootID then it is not the root.
+	current := commentsMap[leafID]
+	for current.ParentID != 0 {
+		// Check if next parent in the tree is the rootID.
+		if current.ParentID == rootID {
+			return true
+		}
+		leafID = current.ParentID
+		current = commentsMap[leafID]
+	}
+	return false
+}
+
+// latestAuthorUpdate gets the latest author update on a record, if
+// the record has no author update it returns nil.
+func latestAuthorUpdate(token []byte, cs []comments.Comment) (*comments.Comment, error) {
+	var latestAuthorUpdate comments.Comment
+	for _, c := range cs {
+		if c.ExtraDataHint != pi.ProposalUpdateHint {
+			continue
+		}
+		if c.Timestamp > latestAuthorUpdate.Timestamp {
+			latestAuthorUpdate = c
+		}
+	}
+	return &latestAuthorUpdate, nil
+}
+
+// recordAuthor returns the author's userID of the record associated with
+// the provided token.
+func (p *piPlugin) recordAuthor(token []byte) (string, error) {
+	reply, err := p.backend.PluginRead(token, usermd.PluginID,
+		usermd.CmdAuthor, "")
+	if err != nil {
+		return "", err
+	}
+	var ar usermd.AuthorReply
+	err = json.Unmarshal([]byte(reply), &ar)
+	if err != nil {
+		return "", err
+	}
+	return ar.UserID, nil
+}
+
+// commentVoteAllowedOnApprovedProposal verifies that the given comment
+// vote is allowed on the proposal which finished voting, it's vote was
+// approved and is associated with the provided token.
+func (p *piPlugin) commentVoteAllowedOnApprovedProposal(token []byte, payload string, latestAuthorUpdate comments.Comment, cs []comments.Comment) error {
+	// Decode payload
+	var v comments.Vote
+	err := json.Unmarshal([]byte(payload), &v)
+	if err != nil {
+		return err
+	}
+
+	if !isInCommentTree(latestAuthorUpdate.CommentID, v.CommentID, cs) {
+		return backend.PluginError{
+			PluginID:  pi.PluginID,
+			ErrorCode: uint32(pi.ErrorCodeWritesAllowedOnlyOnUpdates),
+		}
+	}
+
+	return nil
+}
+
+// commentNewAllowedOnApprovedProposal verifies that the given new comment
+// is allowed on the proposal which finished voting, it's vote was approved
+// and is associated with the provided token.
+func (p *piPlugin) commentNewAllowedOnApprovedProposal(token []byte, payload string, latestAuthorUpdate comments.Comment, cs []comments.Comment) error {
+	// Decode payload
+	var n comments.New
+	err := json.Unmarshal([]byte(payload), &n)
+	if err != nil {
+		return err
+	}
+
+	// A new comment on an approved proposal must either be an update
+	// from the author (parent ID will be 0) or a reply to the latest
+	// author update.
+	isUpdateReply := isInCommentTree(latestAuthorUpdate.CommentID, n.ParentID, cs)
+	switch {
+	case n.ParentID == 0:
+		// This might be an update from the author.
```

Comment:
You've got 4 levels of nesting inside of this switch statement case.  Write this cleaner.